README.txt
.. contents::

.. Note!
   -----
   
   - code repository
   - bug tracker
   - questions/comments feedback mail


- Code repository: http://svn.syslab.com/svn/syslabcom/Products.PublicJobVacancy
- Questions and comments to info (at) syslab (dot) com
- Report bugs at http://products.syslab.com/products/Products.PublicJobVacancy



Credits
=======

Copyright European Agency for Health and Safety at Work and Syslab.com
GmbH.

Products.PublicJobVacancy development was funded by the European
Agency for Health and Safety at Work.


License
=======

Products.PublicJobVacancy is licensed under the GNU Lesser Generic
Public License, version 2 or later and EUPL version 1.1 only. The
complete license texts can be found in docs/LICENSE.GPL and
docs/LICENSE.EUPL.


CHANGES.txt
Products.PublicJobVacancy Changelog
===================================


1.4.2 (unreleased)
------------------

- Nothing changed yet.


1.4.1 (2013-08-21)
------------------

- make deadline field required on PublicJobVacancy type
  [jcerjak]

- specify Products.RichDocument as a dependency and install it automatically
  [jcerjak]

- properly register skin resources
  [jcerjak]

- code cleanup
  [jcerjak]


1.4.0 (2012-09-17)
------------------

- Plone 4


1.3 (unreleased)
----------------

- Added EUPL license (deroiste)


1.2 (2009-05-19)
----------------

- added tests (thomas)

- cleanup (thomas)


1.1 (2009-05-12)
----------------

- Packaged egg (pilz)


1.0 (2008-03-31)
----------------

- Initial port


Products/PublicJobVacancy/tests/base.py
from plone.app.testing import FunctionalTesting
from plone.app.testing import IntegrationTesting
from plone.app.testing import PLONE_FIXTURE
from plone.app.testing import PloneSandboxLayer
from plone.app.testing import applyProfile
from plone.testing import z2


class PublicJobVacancy(PloneSandboxLayer):

    defaultBases = (PLONE_FIXTURE,)

    def setUpZope(self, app, configurationContext):
        import osha.policy.browser
        self.loadZCML('configure.zcml', package=osha.policy.browser)
        import Products.SimpleAttachment
        self.loadZCML('configure.zcml', package=Products.SimpleAttachment)
        import Products.RichDocument
        self.loadZCML('configure.zcml', package=Products.RichDocument)
        import Products.PublicJobVacancy
        self.loadZCML('configure.zcml', package=Products.PublicJobVacancy)

        z2.installProduct(app, 'Products.SimpleAttachment')
        z2.installProduct(app, 'Products.RichDocument')
        z2.installProduct(app, 'Products.PublicJobVacancy')

    def setUpPloneSite(self, portal):
        # Needed to make skins work
        applyProfile(portal, 'Products.CMFPlone:plone')

        applyProfile(portal, 'Products.SimpleAttachment:default')
        applyProfile(portal, 'Products.RichDocument:default')
        applyProfile(portal, 'Products.PublicJobVacancy:default')

    def tearDownZope(self, app):
        z2.uninstallProduct(app, 'Products.SimpleAttachment')
        z2.uninstallProduct(app, 'Products.RichDocument')
        z2.uninstallProduct(app, 'Products.PublicJobVacancy')


PUBLICJOBVACANCY_FIXTURE = PublicJobVacancy()
INTEGRATION_TESTING = IntegrationTesting(
    bases=(PUBLICJOBVACANCY_FIXTURE,),
    name="PublicJobVacancy:Integration")
FUNCTIONAL_TESTING = FunctionalTesting(
    bases=(PUBLICJOBVACANCY_FIXTURE,),
    name="PublicJobVacancy:Functional")


Products/PublicJobVacancy/tests/test_content.py
from plone.app.testing import setRoles
from plone.app.testing import TEST_USER_ID
from plone.app.testing import TEST_USER_NAME
from plone.app.testing import login
from Products.PublicJobVacancy.tests.base import INTEGRATION_TESTING

import unittest2 as unittest


class TestPublicJobVacancy(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']

        # Login as manager
        setRoles(self.portal, TEST_USER_ID, ['Manager'])
        login(self.portal, TEST_USER_NAME)

    def test_create(self):
        """Ensure that we can create a PublicJobVacancy without errors"""

        self.portal.invokeFactory(
            'PublicJobVacancy', 'job-vacancy', title=u"Public Job vacancy")


def test_suite():
    """This sets up a test suite that actually runs the tests in
    the class(es) above.
    """
    return unittest.defaultTestLoader.loadTestsFromName(__name__)


Products/PublicJobVacancy/tests/test_doctest.py
import doctest
import unittest2 as unittest

from Products.PublicJobVacancy.tests.base import FUNCTIONAL_TESTING
from plone.testing import layered

OPTIONFLAGS = (doctest.REPORT_ONLY_FIRST_FAILURE |
               doctest.ELLIPSIS |
               doctest.NORMALIZE_WHITESPACE)


def test_suite():
    suite  = unittest.TestSuite()
    suite.addTests([
        layered(
                # Demonstrate the main content types
                doctest.DocFileSuite(
                    "../README.txt",
                    optionflags=OPTIONFLAGS),
                layer=FUNCTIONAL_TESTING),
            ])
    return suite


Products/PublicJobVacancy/config.py
# -*- coding: utf-8 -*-
#
# File: PublicJobVacancy.py
#
# Copyright (c) 2007 by []
# Generator: ArchGenXML Version 2.0-beta1
#            http://plone.org/products/archgenxml
#
# GNU General Public License (GPL)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#

__author__ = """Syslab.com GmbH <info@syslab.com>"""
__docformat__ = 'plaintext'


# Product configuration.
#
# The contents of this module will be imported into __init__.py, the
# workflow configuration and every content type module.
#
# If you wish to perform custom configuration, you may put a file
# AppConfig.py in your product's root directory. The items in there
# will be included (by importing) in this file if found.

from Products.CMFCore.permissions import setDefaultRoles

PROJECTNAME = "PublicJobVacancy"

# Permissions
DEFAULT_ADD_CONTENT_PERMISSION = "Add portal content"
setDefaultRoles(DEFAULT_ADD_CONTENT_PERMISSION, ('Manager', 'Owner'))

product_globals = globals()

# Dependencies of Products to be installed by quick-installer
# override in custom configuration
DEPENDENCIES = []

# Dependend products - not quick-installed - used in testcase
# override in custom configuration
PRODUCT_DEPENDENCIES = []


# Load custom configuration not managed by archgenxml
try:
    from Products.PublicJobVacancy.AppConfig import *
except ImportError:
    pass


Products/PublicJobVacancy/__init__.py
# -*- coding: utf-8 -*-
#
# File: PublicJobVacancy.py
#
# Copyright (c) 2007 by []
# Generator: ArchGenXML Version 2.0-beta1
#            http://plone.org/products/archgenxml
#
# GNU General Public License (GPL)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#

__author__ = """Syslab.com GmbH <info@syslab.com>"""
__docformat__ = 'plaintext'


# There are three ways to inject custom code here:
#
#   - To set global configuration variables, create a file AppConfig.py.
#       This will be imported in config.py, which in turn is imported in
#       each generated class and in this file.
#   - To perform custom initialisation after types have been registered,
#       use the protected code section at the bottom of initialize().

import logging
logger = logging.getLogger('PublicJobVacancy')
logger.debug('Installing Product')

import os
import os.path
from Globals import package_home
import Products.CMFPlone.interfaces
from Products.Archetypes import listTypes
from Products.Archetypes.atapi import *
from Products.Archetypes.utils import capitalize
from Products.CMFCore import DirectoryView
from Products.CMFCore import permissions as cmfpermissions
from Products.CMFCore import utils as cmfutils
from Products.CMFPlone.utils import ToolInit
from config import *


def initialize(context):
    """initialize product (called by zope)"""
    ##code-section custom-init-top #fill in your manual code here
    ##/code-section custom-init-top

    # imports packages and types for registration

    import PublicJobVacancy

    # Initialize portal content
    content_types, constructors, ftis = process_types(
        listTypes(PROJECTNAME),
        PROJECTNAME)

    cmfutils.ContentInit(
        PROJECTNAME + ' Content',
        content_types=content_types,
        permission=DEFAULT_ADD_CONTENT_PERMISSION,
        extra_constructors=constructors,
        fti=ftis,
        ).initialize(context)


from zope.i18nmessageid import MessageFactory
JobMessageFactory = MessageFactory('PublicJobVacancy')


Products/PublicJobVacancy/browser/listing.py
from Acquisition import aq_base
from DateTime import DateTime
from DocumentTemplate import sequence
from Products.CMFCore.utils import getToolByName
from Products.Five.browser import BrowserView
from Products.PublicJobVacancy.browser.interfaces import IJobsListing
from zope.interface import implements


class JobsSortedListing(BrowserView):
    implements(IJobsListing)

    def getAdditionalText(self):
        context = self.context
        text = ''
        if getattr(aq_base(context), 'content', None):
            content = getattr(context, 'content')
            if hasattr(content, 'getText') and callable(content.getText):
                text = content.getText()
        return text

    def results(self):
        context = self.context
        portal_catalog = getToolByName(context, 'portal_catalog')
        res = portal_catalog({
            'portal_type': 'PublicJobVacancy',
            'sort_on': 'effective',
            'Language': 'all',
            'sort_order': 'reverse'
        })
        effRes = list()
        cres = list()
        ores = list()
        sres = list()
        ares = list()
        date = DateTime()

        for r in res:
            try:
                o = r.getObject()
            except:
                continue
            pastEffective = (r.effective is None or r.effective == ''
                             or r.effective <= date)
            beforeExpiration = (r.expires is None or r.expires == ''
                                 or r.expires >= date)
            if r.effective is None and r.expires is None:
                continue

            if pastEffective:
                if beforeExpiration:
                    effRes.append(r)
                    if o.deadline > date:
                        cres.append(r)
                    else:
                        if not o.getShortlisted():
                            ores.append(r)
                        else:
                            sres.append(r)
                else:
                    ares.append(r)
        cres = sequence.sort(cres, (('effective', 'cmp', 'desc'),
                                    ('CreationDate', 'cmp', 'desc')))
        ores = sequence.sort(ores, (('effective', 'cmp', 'desc'),
                                    ('CreationDate', 'cmp', 'desc')))
        sres = sequence.sort(sres, (('effective', 'cmp', 'desc'),
                                    ('CreationDate', 'cmp', 'desc')))
        return dict(current=cres,
                    ongoing=ores,
                    shortlisted=sres,
                    archive=ares)


Products/PublicJobVacancy/PublicJobVacancy.py
# -*- coding: utf-8 -*-
#
# File: PublicJobVacancy.py
##
# GNU General Public License (GPL)

__author__ = """Syslab.com GmbH <info@syslab.com>"""
__docformat__ = 'plaintext'

from AccessControl import ClassSecurityInfo
from Products.Archetypes.atapi import *

try:
    from Products.LinguaPlone.public import *
except ImportError:
    HAS_LINGUAPLONE = False
else:
    HAS_LINGUAPLONE = True

from zope.interface import implements
from interfaces import IPublicJobVacancy

from Products.CMFDynamicViewFTI.browserdefault import BrowserDefaultMixin

from Products.PublicJobVacancy.config import *

from Products.RichDocument.content.richdocument import RichDocument
from Products.RichDocument.content.richdocument import \
    RichDocumentSchema as BaseSchema
from Products.ATContentTypes.content.schemata import finalizeATCTSchema

if HAS_LINGUAPLONE:
    from Products.LinguaPlone.public import registerType
else:
    from Products.Archetypes.public import registerType

from Products.PublicJobVacancy import JobMessageFactory as _

schema = Schema((

    StringField(
        name='author',
        widget=StringWidget(
            size=80,
            label=_(u'job_author_label', default=u"Author"),
            description=_(u'job_author_description',
                          u"The author issuing the job"),
        )
    ),

    DateTimeField(
        name='deadline',
        required=True,
        widget=CalendarWidget(
            size="CalendarWidget",
            label=_(u'job_deadline_label', default="Deadline"),
            description=_(u'job_deadline_description',
                          default=u"The deadline for this job vacancy"),
        ),
        languageIndependent=True
    ),

    BooleanField(
        name='shortlisted',
        widget=BooleanField._properties['widget'](
            label=_(u'job_shortlisted_label', default=u"Shortlisted"),
            description=_(u'job_shortlisted_description',
                          default=u"Has this job vacancy been shortlisted?"),
        ),
        languageIndependent=True
    ),

),
)


PublicJobVacancy_schema = BaseSchema.copy() + schema.copy()
finalizeATCTSchema(PublicJobVacancy_schema)


class PublicJobVacancy(BrowserDefaultMixin, BaseContent, RichDocument):
    """Public Job Vacancy type"""

    security = ClassSecurityInfo()
    implements(IPublicJobVacancy)
    meta_type = 'PublicJobVacancy'
    _at_rename_after_creation = True

    schema = PublicJobVacancy_schema


registerType(PublicJobVacancy, PROJECTNAME)


Products/PublicJobVacancy/setuphandlers.py
# -*- coding: utf-8 -*-
#
# File: setuphandlers.py
#
# Copyright (c) 2007 by []
# Generator: ArchGenXML Version 2.0-beta1
#            http://plone.org/products/archgenxml
#
# GNU General Public License (GPL)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#

__author__ = """Syslab.com GmbH <info@syslab.com>"""
__docformat__ = 'plaintext'


import logging
logger = logging.getLogger('PublicJobVacancy: setuphandlers')
from config import DEPENDENCIES
from Products.CMFCore.utils import getToolByName
from Products.SimpleAttachment.setuphandlers import \
    registerImagesFormControllerActions
from Products.SimpleAttachment.setuphandlers import \
    registerAttachmentsFormControllerActions


def installGSDependencies(context):
    """Install dependend profiles."""

    # XXX Hacky, but works for now. has to be refactored as soon as generic
    # setup allows a more flexible way to handle dependencies.

    dependencies = []
    if not dependencies:
        return

    site = context.getSite()
    setup_tool = getToolByName(site, 'portal_setup')
    for dependency in dependencies:
        if dependency.find(':') == -1:
            dependency += ':default'
        old_context = setup_tool.getImportContextID()
        setup_tool.setImportContext('profile-%s' % dependency)
        importsteps = setup_tool.getImportStepRegistry().sortSteps()
        excludes = [
            u'PublicJobVacancy-QI-dependencies',
            u'PublicJobVacancy-GS-dependencies'
        ]
        importsteps = [s for s in importsteps if s not in excludes]
        for step in importsteps:
            setup_tool.runImportStep(step)  # purging flag here?
        setup_tool.setImportContext(old_context)

    # re-run some steps to be sure the current profile applies as expected
    importsteps = setup_tool.getImportStepRegistry().sortSteps()
    filter = [
        u'typeinfo',
        u'workflow',
        u'membranetool',
        u'factorytool',
        u'content_type_registry',
        u'membrane-sitemanager'
    ]
    importsteps = [s for s in importsteps if s in filter]
    for step in importsteps:
        setup_tool.runImportStep(step)  # purging flag here?


def installQIDependencies(context):
    """This is for old-style products using QuickInstaller"""
    site = context.getSite()
    qi = getToolByName(site, 'portal_quickinstaller')
    for dependency in DEPENDENCIES:
        if qi.isProductInstalled(dependency):
            logger.info("Re-Installing dependency %s:" % dependency)
            qi.reinstallProducts([dependency])
        else:
            logger.info("Installing dependency %s:" % dependency)
            qi.installProducts([dependency])


def installAttachmentSupport(context):
    """ install attachment handlers """
    site = context.getSite()
    # Set up form controller actions for the widgets to work
    registerAttachmentsFormControllerActions(
        site, contentType='PublicJobVacancy', template='base_edit')
    registerImagesFormControllerActions(
        site, contentType='PublicJobVacancy', template='base_edit')

    # Register form controller actions for LinguaPlone translate_item
    registerAttachmentsFormControllerActions(
        site, contentType='PublicJobVacancy', template='translate_item')
    registerImagesFormControllerActions(
        site, contentType='PublicJobVacancy', template='translate_item')
